pkg/operator/ceph/cluster/mgr/drain.go

```diff
@@ -0,0 +1,111 @@
+*/
+
+package mgr
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	policyv1 "k8s.io/api/policy/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	mgrPDBName = "rook-ceph-mgr-pdb"
+)
+
+func (c *Cluster) reconcileMgrPDB() error {
+	op, err := c.createOrUpdateMgrPDB(1)
+	if err != nil {
+		return errors.Wrapf(err, "failed to reconcile mgr pdb on op %q", op)
+	}
+	return nil
+}
+
+func (c *Cluster) createOrUpdateMgrPDB(maxUnavailable int32) (controllerutil.OperationResult, error) {
+	usePDBV1Beta1, err := k8sutil.UsePDBV1Beta1Version(c.context.Clientset)
+	if err != nil {
+		return controllerutil.OperationResultNone, errors.Wrap(err, "failed to fetch pdb version")
+	}
+	objectMeta := metav1.ObjectMeta{
+		Name:      mgrPDBName,
+		Namespace: c.clusterInfo.Namespace,
+	}
+	selector := &metav1.LabelSelector{
+		MatchLabels: map[string]string{k8sutil.AppAttr: AppName},
+	}
+	if usePDBV1Beta1 {
+		pdb := &policyv1beta1.PodDisruptionBudget{
+			ObjectMeta: objectMeta,
+		}
+		mutateFunc := func() error {
+			pdb.Spec = policyv1beta1.PodDisruptionBudgetSpec{
+				Selector:       selector,
+				MaxUnavailable: &intstr.IntOrString{IntVal: maxUnavailable},
+			}
+			return nil
+		}
+		return controllerutil.CreateOrUpdate(context.TODO(), c.context.Client, pdb, mutateFunc)
+	}
+	pdb := &policyv1.PodDisruptionBudget{
+		ObjectMeta: objectMeta,
+	}
+	mutateFunc := func() error {
+		pdb.Spec = policyv1.PodDisruptionBudgetSpec{
+			Selector:       selector,
+			MaxUnavailable: &intstr.IntOrString{IntVal: maxUnavailable},
+		}
+		return nil
+	}
+	return controllerutil.CreateOrUpdate(context.TODO(), c.context.Client, pdb, mutateFunc)
+}
+
+func (c *Cluster) deleteMgrPDB() {
+	pdbRequest := types.NamespacedName{Name: mgrPDBName, Namespace: c.clusterInfo.Namespace}
+	usePDBV1Beta1, err := k8sutil.UsePDBV1Beta1Version(c.context.Clientset)
+	if err != nil {
+		return
+	}
+	if usePDBV1Beta1 {
+		mgrPDB := &policyv1beta1.PodDisruptionBudget{}
+		err := c.context.Client.Get(context.TODO(), pdbRequest, mgrPDB)
+		if err != nil {
+			return
+		}
+		logger.Debugf("ensuring the mgr pdb %q is deleted", mgrPDBName)
+		err = c.context.Client.Delete(context.TODO(), mgrPDB)
+		if err != nil && !kerrors.IsNotFound(err) {
+			logger.Errorf(err.Error(), "failed to delete mgr pdb %q", mgrPDBName)
+		}
+	}
+	mgrPDB := &policyv1.PodDisruptionBudget{}
+	err = c.context.Client.Get(context.TODO(), pdbRequest, mgrPDB)
+	if err != nil {
+		return
+	}
+	logger.Infof("deleting the mgr pdb %q with maxUnavailable=1", mgrPDBName)
+	err = c.context.Client.Delete(context.TODO(), mgrPDB)
+	if err != nil && !kerrors.IsNotFound(err) {
+		logger.Errorf(err.Error(), "failed to delete mgr pdb %q", mgrPDBName)
```

Comment:
any particular reason why `delete` error is not handled ?

Reply:
if the mgr count is 1, and the MGR-PDB doesn't exist then
I guess no need to return the error and stop the mgr reconcile, 

But we can check for PDB exist or not and then we can handle the errors, @travisn any more views for ignoring errors, or I should have to do that.

Reply:
IMO the delete is not critical to the reconcile so if we just log the error it is fine. If it should be deleted and the error was intermittent, it will anyway be cleaned up at the next reconcile.